fs/readdircount_linux_amd64.go

```diff
@@ -0,0 +1,62 @@
+// +build linux,amd64
+
+package fs
+
+import (
+	"fmt"
+	"os"
+	"unsafe"
+
+	"syscall"
+)
+
+func countDirEntries(buf []byte, n int) int {
+	count := 0
+	buf = buf[:n]
+	for len(buf) > 0 {
+		// see man page getdents(2) for struct linux_dirent64
+		reclenOffset := unsafe.Offsetof(syscall.Dirent{}.Reclen)
+		reclen := *(*uint16)(unsafe.Pointer(&buf[reclenOffset]))
+
+		inoOffset := unsafe.Offsetof(syscall.Dirent{}.Ino)
+		ino := *(*uint64)(unsafe.Pointer(&buf[inoOffset]))
```

Comment:
Without unsafe.Pointer, I get:
```
./readdircount_linux_amd64.go:19: cannot convert &buf[reclenOffset] (type *byte) to type *uint16
```
I think the situation is different than in your link because `uint64(b[0])` is not reading the pointer but the value at index 0.

We need to specify the uint16 / uint64  so that it takes the correct amount of bytes from the array to form the integer. Otherwise, it would just take one byte because `buf` is a `[]byte`